lib/logger.js

```diff
@@ -522,83 +614,104 @@ class Logger extends EventEmitter {
-        }
-
-        this.emit('cleared', {
-          message: ALL_CLEAR_SENT
-        })
-      })
-      .catch((error) => {
-        const code = error.response
-          ? error.response.status
-          : error.code // timeouts will populate this
+    const data = stringify({e: 'ls', ls: buffer})
 
-        const retrying = ERROR_CODES_TO_RETRY.has(code)
-        const err = new Error('An error occured while sending logs to the cloud.')
+    this._getSendPayload(data, (error, payload) => {
+      if (error) {
+        const err = new Error('Error gzipping data')
         err.meta = {
-          actual: error.message
-        , code
-        , firstLine
-        , lastLine
-        , retrying
-        , attempts: ++this[kAttempts]
+          message: 'Will attempt to send data uncompressed'
+        , error
         }
+        // We will still send, but without compression
+        delete config.headers['Content-Encoding']
+        config.data = data
         this.emit('error', err)
+      } else {
+        config.data = payload
+      }
 
-        if (retrying) {
-          this[kIsLoggingBackedOff] = true
-          this[kBackoffMs] = backoffWithJitter(
-            this.baseBackoffMs
-          , this.maxBackoffMs
-          , this[kBackoffMs]
-          )
-          setTimeout(() => {
-            this.send(false)
-          }, this[kBackoffMs])
-          return
-        }
-
-        // User-level errors will be discarded since they will never succeed
-        this[kIsSending] = false
-        this[kTotalLinesReady] -= buffer.length
-        this[kAttempts] = 0
-        this[kReadyToSend].shift()
-        buffer.length = 0
-        if (this[kReadyToSend].length) {
-          this.send()
-          return
-        }
+      axios(config)
+        .then((response) => {
+          // We have a 200-level success code.
+          const totalLinesSent = buffer.length
+          this[kIsLoggingBackedOff] = false
+          this[kAttempts] = 0
+          this[kIsSending] = false
+          this[kTotalLinesReady] -= totalLinesSent
+          this[kReadyToSend].shift()
+
+          // Assist GC by killing the buffer and removing it from readyToSend
+          buffer.length = 0
+
+          this.emit('send', {
+            httpStatus: response.status
+          , firstLine
+          , lastLine
+          , totalLinesSent
+          , totalLinesReady: this[kTotalLinesReady]
+          , bufferCount: this[kReadyToSend].length
+          })
+
+          if (this[kReadyToSend].length) {
+            // Continue to send any backed up payloads that have accumulated
+            this.send()
+            return
+          }
 
-        this.emit('cleared', {
-          message: ALL_CLEAR_SENT
+          this.emit('cleared', {
+            message: ALL_CLEAR_SENT
+          })
         })
-      })
+        .catch((error) => {
+          // Allow the microtask queue to unwind in case it's a Promise rejection
+          process.nextTick(() => {
```

Comment:
To be clear, all this code isn't "new", it's just indented a block.  If a promise inside of axios has thrown, we need to unwind the stack.